data_model/derive/src/id.rs

```diff
@@ -1,16 +1,109 @@
 #![allow(clippy::str_to_string, clippy::mixed_read_write_in_expression)]
 
-use manyhow::{bail, Result};
+use darling::{FromAttributes, FromDeriveInput, FromField};
+use iroha_macro_utils::Emitter;
+use manyhow::emit;
 use proc_macro2::TokenStream;
-use quote::quote;
+use quote::{quote, ToTokens};
 use syn2::parse_quote;
 
-pub fn impl_id(input: &syn2::ItemStruct) -> Result<TokenStream> {
+mod kw {
+    syn2::custom_keyword!(transparent);
+}
+
+enum IdAttr {
+    Missing,
+    Normal,
+    Transparent,
+}
+
+impl FromAttributes for IdAttr {
+    fn from_attributes(attrs: &[syn2::Attribute]) -> darling::Result<Self> {
+        let mut accumulator = darling::error::Accumulator::default();
+        let attrs = attrs
+            .iter()
+            .filter(|v| v.path().is_ident("id"))
+            .collect::<Vec<_>>();
+        let attr = match attrs.as_slice() {
+            [] => {
+                return accumulator.finish_with(IdAttr::Missing);
+            }
+            [attr] => attr,
+            [attr, ref tail @ ..] => {
+                accumulator.push(
+                    darling::Error::custom("Only one `#[id]` attribute is allowed!").with_span(
+                        &tail
+                            .iter()
+                            .map(syn2::spanned::Spanned::span)
+                            .reduce(|a, b| a.join(b).unwrap())
+                            .unwrap(),
+                    ),
+                );
+                attr
+            }
+        };
+
+        let result = match &attr.meta {
+            syn2::Meta::Path(_) => IdAttr::Normal,
+            syn2::Meta::List(list) if list.parse_args::<kw::transparent>().is_ok() => {
+                IdAttr::Transparent
+            }
+            _ => {
+                accumulator.push(
+                    darling::Error::custom("Expected `#[id]` or `#[id(transparent)]`")
+                        .with_span(&attr),
+                );
+                IdAttr::Normal
+            }
+        };
+
+        accumulator.finish_with(result)
+    }
```

Comment:
Isn't it a copy-paste of code that originally was from ffi?

Reply:
Yeah, it does have some similarities.. I would need to add some other functions, but I think it can be generalized

Reply:
Done with the `find_single_attr_opt` function